pkg/netceptor/ping_test.go

```diff
@@ -0,0 +1,251 @@
+	err        error
+	times      int
+	returnType string
+}
+
+// TestCreatePing tests CreatePing inside ping.go.
+func TestCreatePing(t *testing.T) {
+	ctrl, mockNetceptor, mockPacketConn := setupTest(t)
+
+	pingTestCases := []struct {
+		name               string
+		pingTarget         string
+		pingHopsToLive     byte
+		returnListenPacket listenPacketReturn
+		returnWriteTo      writeToReturn
+		returnContext      contextReturn
+		returnReadFrom     readFromReturn
+		expectedDuration   int
+		expectedRemote     string
+		expectedError      error
+	}{
+		{"NetceptorShutdown Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "doAndReturn"}, readFromReturn{0, nil, nil, 1, "return"}, 0, "", errors.New("netceptor shutdown")},
+		{"ListenSubscribeUnreachable Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, nil, nil, 1, "return"}, 0, "", errors.New("test")},
+		{"CreatePing Success", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, &netceptor.Addr{}, nil, 1, "return"}, 0, ":", nil},
+		{"ListenPacket Error", "target", byte(1), listenPacketReturn{nil, errors.New("Catch ListenPacket error"), 1, "return"}, writeToReturn{0, nil, 0, "return"}, contextReturn{context.Background(), 0, "return"}, readFromReturn{0, &netceptor.Addr{}, nil, 0, "return"}, 0, "", errors.New("Catch ListenPacket error")},
+		{"ReadFrom Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, nil, errors.New("ReadFrom error"), 1, "return"}, 0, "", errors.New("ReadFrom error")},
+		{"WriteTo Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, errors.New("WriteTo error"), 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, nil, nil, 1, "return"}, 0, "", errors.New("WriteTo error")},
+		{"Timeout Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, nil, nil, 1, "do"}, 0, "", errors.New("timeout")},
+		{"User Cancel Error", "target", byte(1), listenPacketReturn{mockPacketConn, nil, 1, "return"}, writeToReturn{0, nil, 1, "return"}, contextReturn{context.Background(), 2, "return"}, readFromReturn{0, nil, nil, 1, "doAndReturn"}, 0, "", errors.New("user cancelled")},
+	}
+
+	for _, testCase := range pingTestCases {
+		ctx := context.Background()
+		t.Run(testCase.name, func(t *testing.T) {
+			// Set up the mock behaviours
+			if testCase.name != "ListenSubscribeUnreachable Error" {
```

Comment:
I see why you have branched so many times here, good job for working through this logic. 

Perhaps we could look at alternative ways of doing this? 
First thing that comes to mind is function pointers, perhaps `pingTestCases` could have two function pointer fields, `setUp` and `checkResults`. 
This way each test case could have a unique set up and results check without this if conditions being needed. WDYT?